File: src/main/java/org/jcodec/containers/mp4/boxes/TimecodeMediaInfoBox.java
package org.jcodec.containers.mp4.boxes;

import org.jcodec.common.io.NIOUtils;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class TimecodeMediaInfoBox extends FullBox {

    private short font;
    private short face;
    private short size;
    private short[] color;
    private short[] bgcolor;
    private String name;

    public static String fourcc() {
        return "tcmi";
    }

    public static TimecodeMediaInfoBox createTimecodeMediaInfoBox(short font, short face, short size, short[] color,
            short[] bgcolor, String name) {
        TimecodeMediaInfoBox box = new TimecodeMediaInfoBox(new Header(fourcc()));
        box.font = font;
        box.face = face;
        box.size = size;
        box.color = color;
        box.bgcolor = bgcolor;
        box.name = name;
        return box;
    }

    public TimecodeMediaInfoBox(Header atom) {
        super(atom);
        this.color = new short[3];
        this.bgcolor = new short[3];
    }

    @Override
    public void parse(ByteBuffer input) {
        super.parse(input);
        font = input.getShort();
        face = input.getShort();
        size = input.getShort();
        input.getShort();
        color[0] = input.getShort();
        color[1] = input.getShort();
        color[2] = input.getShort();
        bgcolor[0] = input.getShort();
        bgcolor[1] = input.getShort();
        bgcolor[2] = input.getShort();
        name = NIOUtils.readPascalString(input);
    }

    @Override
    protected void doWrite(ByteBuffer out) {
        super.doWrite(out);
        out.putShort(font);
        out.putShort(face);
        out.putShort(size);
        out.putShort((short) 0);
        out.putShort(color[0]);
        out.putShort(color[1]);
        out.putShort(color[2]);
        out.putShort(bgcolor[0]);
        out.putShort(bgcolor[1]);
        out.putShort(bgcolor[2]);
        NIOUtils.writePascalString(out, name);
    }

    @Override
    public int estimateSize() {
        return 32 + 1 + NIOUtils.asciiString(name).length;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/TimecodeMediaInfoBox.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.common.io.NIOUtils
import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class TimecodeMediaInfoBox(atom: Header) : FullBox(atom) {
    private var font: Short = 0
    private var face: Short = 0
    private var size: Short = 0
    private var color: ShortArray
    private var bgcolor: ShortArray
    private var name: String? = null
    override fun parse(input: ByteBuffer) {
        super.parse(input)
        font = input.short
        face = input.short
        size = input.short
        input.short
        color[0] = input.short
        color[1] = input.short
        color[2] = input.short
        bgcolor[0] = input.short
        bgcolor[1] = input.short
        bgcolor[2] = input.short
        name = NIOUtils.readPascalString(input)
    }

    override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putShort(font)
        out.putShort(face)
        out.putShort(size)
        out.putShort(0.toShort())
        out.putShort(color[0])
        out.putShort(color[1])
        out.putShort(color[2])
        out.putShort(bgcolor[0])
        out.putShort(bgcolor[1])
        out.putShort(bgcolor[2])
        NIOUtils.writePascalString(out, name)
    }

    override fun estimateSize(): Int {
        return 32 + 1 + NIOUtils.asciiString(name).size
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "tcmi"
        }

        @JvmStatic
        fun createTimecodeMediaInfoBox(font: Short, face: Short, size: Short, color: ShortArray,
                                       bgcolor: ShortArray, name: String?): TimecodeMediaInfoBox {
            val box = TimecodeMediaInfoBox(Header(fourcc()))
            box.font = font
            box.face = face
            box.size = size
            box.color = color
            box.bgcolor = bgcolor
            box.name = name
            return box
        }
    }

    init {
        color = ShortArray(3)
        bgcolor = ShortArray(3)
    }
}